client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamFactoryImpl.java

```diff
@@ -88,4 +96,37 @@ public SegmentInputStream createInputStreamForSegment(Segment segment, Delegatio
         async.getConnection();
         return new SegmentInputStreamImpl(async, startOffset);
     }
+
+    protected DelegationTokenProvider getDelegationTokenProvider(Segment segment) {
+        log.debug("getDelegationTokenProvider, scopedName: {}, cacheInfo: {}", segment.getScopedName(), tokenProviderCache.printCache());
+        DelegationTokenProvider tokenProvider = tokenProviderCache.get(segment.getScopedName());
```

Comment:
Rather than using a per-segment cache it makes more sense to have a per stream cache. It is possible to create a DelegationTokenProvider given a stream. If we take advantage of this we will be able to consolidate entries in the cache.